File: TelerikUIExamples/TelerikUIExamples/CustomAnimationAreaChart.m
//
//  CustomAnimationAreaChart.m
//  TelerikUIExamples
//
//  Copyright (c) 2013 Telerik. All rights reserved.
//

#import "CustomAnimationAreaChart.h"
#import <TelerikUI/TelerikUI.h>

@interface CustomAnimationAreaChart ()

@end

@implementation CustomAnimationAreaChart
{
    TKChart *_chart;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _chart = [[TKChart alloc] initWithFrame:[self exampleBounds]];
    _chart.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _chart.allowAnimations = YES;
    _chart.delegate = self;
    [self.view addSubview:_chart];
    
    NSMutableArray *points = [[NSMutableArray alloc] init];
    for (int i = 0; i<10; i++) {
        [points addObject:[[TKChartDataPoint alloc] initWithX:@(i) Y:@(arc4random() % 100)]];
    }
    
    TKChartAreaSeries *areaSeries = [[TKChartAreaSeries alloc] initWithItems:points];
    areaSeries.selectionMode = TKChartSeriesSelectionModeSeries;
    [_chart addSeries:areaSeries];
    
    [[UIDevice currentDevice] beginGeneratingDeviceOrientationNotifications];
     [[NSNotificationCenter defaultCenter] addObserver:self
                                              selector:@selector(deviceOrientationDidChange:)
                                                  name:UIDeviceOrientationDidChangeNotification
                                                object:nil];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    [[UIDevice currentDevice] endGeneratingDeviceOrientationNotifications];
}

- (void)deviceOrientationDidChange:(NSNotification *)notification
{
    [_chart animate];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark TKChartDelegate

- (CAAnimation *)chart:(TKChart *)chart animationForSeries:(TKChartSeries *)series withState:(TKChartSeriesRenderState *)state inRect:(CGRect)rect
{
    CFTimeInterval duration = .5;
    NSMutableArray *animations = [[NSMutableArray alloc] init];
    
    for (int i = 0; i<state.points.count; i++) {
        NSString *keyPath = [NSString stringWithFormat:@"seriesRenderStates.%lu.points.%d.y", (unsigned long)series.index, i];
        TKChartVisualPoint *point = [state.points objectAtIndex:i];
        CGFloat oldY = rect.size.height;
        
        CGFloat half = oldY + (point.y - oldY)/2.;
        CAKeyframeAnimation *a = [CAKeyframeAnimation animationWithKeyPath:keyPath];
        a.keyTimes = @[ @0, @0, @1 ];
        a.values = @[ @(oldY), @(half), @(point.y) ];
        a.duration = duration;
        a.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
        [animations addObject:a];
    }
    
    CAAnimationGroup *group = [[CAAnimationGroup alloc] init];
    group.duration = duration;
    group.animations = animations;
    return group;
}

@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/CustomAnimationAreaChart.swift
//
//  CustomAnimationAreaChart.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import Foundation
import QuartzCore

class CustomAnimationAreaChart: TKExamplesExampleViewController, TKChartDelegate {
    
    let chart = TKChart()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        chart.frame = self.view.bounds
        chart.autoresizingMask = UIViewAutoresizing(rawValue: UIViewAutoresizing.FlexibleWidth.rawValue | UIViewAutoresizing.FlexibleHeight.rawValue)
        chart.allowAnimations = true
        chart.delegate = self
        self.view.addSubview(chart)

        var points = [TKChartDataPoint]()
        for i in 0..<10 {
            points.append(TKChartDataPoint(x: i, y: Int(arc4random() % 100)))
        }
        
        let areaSeries = TKChartAreaSeries(items: points)
        areaSeries.selectionMode = TKChartSeriesSelectionMode.Series
        chart.addSeries(areaSeries)
        
        UIDevice.currentDevice().beginGeneratingDeviceOrientationNotifications()
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "deviceOrientationDidChange:",
            name: UIDeviceOrientationDidChangeNotification, object: nil)
    }
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
        UIDevice.currentDevice().endGeneratingDeviceOrientationNotifications()
    }
    
    func deviceOrientationDidChange(notification: NSNotification) {
        chart.animate()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //MARK: - TKChartDelegate
    
    func chart(chart: TKChart, animationForSeries series: TKChartSeries, withState state: TKChartSeriesRenderState, inRect rect: CGRect) -> CAAnimation? {
        let duration = 0.5
        var animations = [CAAnimation]()
        for i in 0..<state.points.count() {
            let keyPath = "seriesRenderStates.\(series.index).points.\(i).y"
            let point = state.points.objectAtIndex(i) as! TKChartVisualPoint
            let oldY = rect.size.height
            
            let half = oldY + (point.y - oldY)/2 as CGFloat
            let a = CAKeyframeAnimation(keyPath: keyPath)
            a.keyTimes = [0, 0, 1]
            a.values = [oldY, half, point.y]
            a.duration = duration
            a.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
            animations.append(a)
        }
        
        let group = CAAnimationGroup()
        group.duration = duration
        group.animations = animations
        return group
    }
}
